src/test/node-modules-layout-maker_test.ts

```diff
@@ -0,0 +1,303 @@
+/**
+ * @license
+ * Copyright 2021 Google LLC
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+import {assert} from '@esm-bundle/chai';
+import {
+  NodeModulesLayoutMaker,
+  DependencyGraph,
+  NodeModulesDirectory,
+  PackageDependencies,
+} from '../typescript-worker/node-modules-layout-maker.js';
+
+suite('NodeModulesLayoutMaker', () => {
+  const checkLayout = (
+    rootDeps: PackageDependencies,
+    depGraph: DependencyGraph,
+    expected: NodeModulesDirectory
+  ) => {
+    const layoutMaker = new NodeModulesLayoutMaker();
+    const actual = layoutMaker.layout(rootDeps, depGraph);
+    assert.deepEqual(actual, expected);
+  };
+
+  test('no dependencies', () => {
+    // ROOT
+    const rootDeps: PackageDependencies = {};
+    const depGraph: DependencyGraph = {};
+    const expected: NodeModulesDirectory = {};
+    checkLayout(rootDeps, depGraph, expected);
+  });
+
+  test('only direct dependencies', () => {
+    //    ROOT
+    //     /\
+    //    /  \
+    //   v    v
+    //  A1    B2
+    const rootDeps: PackageDependencies = {
+      a: '1',
+      b: '2',
+    };
+    const depGraph: DependencyGraph = {};
+    // ROOT
+    // ├── A1
+    // └── B2
+    const expected: NodeModulesDirectory = {
+      a: {version: '1', nodeModules: {}},
+      b: {version: '2', nodeModules: {}},
+    };
+    checkLayout(rootDeps, depGraph, expected);
+  });
+
+  test('simple chain', () => {
+    //  ROOT
+    //    |
+    //    v
+    //    A1
+    //    |
+    //    v
+    //    B2
+    //    |
+    //    v
+    //    C3
+    const rootDeps: PackageDependencies = {
+      a: '1',
+    };
+    const depGraph: DependencyGraph = {
+      a: {1: {b: '2'}},
+      b: {2: {c: '3'}},
+    };
+    // ROOT
+    // ├── A1
+    // ├── B2
+    // └── C3
+    const expected: NodeModulesDirectory = {
+      a: {version: '1', nodeModules: {}},
+      b: {version: '2', nodeModules: {}},
+      c: {version: '3', nodeModules: {}},
+    };
+    checkLayout(rootDeps, depGraph, expected);
+  });
+
+  test('version conflict between root and branch', () => {
```

Comment:
Is it possible for there to be a conflict with two versions, A1 and A2 directly off ROOT?

Reply:
It is not actually possible because an NPM package can only depend on one version of another package. Dependencies are always declared in a `package.json` `dependencies` field, so it's also not even syntactically possible to declare it. That is, you can't write:

```json
{
  "dependencies": {
    "a": "1.0.0",
    "a": "2.0.0"
  }
}
```

(Well, actually you can write it, but the second "a" clobbers the first one).